pkg/rpcclient/rolemgmt/roles.go

```diff
@@ -0,0 +1,119 @@
+/*
+Package rolemgmt allows to work with the native RoleManagement contract via RPC.
+
+Safe methods are encapsulated into ContractReader structure while Contract provides
+various methods to perform the only RoleManagement state-changing call.
+*/
+package rolemgmt
+
+import (
+	"crypto/elliptic"
+	"fmt"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/native/nativenames"
+	"github.com/nspcc-dev/neo-go/pkg/core/native/noderoles"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neo-go/pkg/neorpc/result"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient/unwrap"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// Invoker is used by ContractReader to call various methods.
+type Invoker interface {
+	Call(contract util.Uint160, operation string, params ...interface{}) (*result.Invoke, error)
+}
+
+// Actor is used by Contract to create and send transactions.
+type Actor interface {
+	Invoker
+
+	MakeCall(contract util.Uint160, method string, params ...interface{}) (*transaction.Transaction, error)
+	MakeUnsignedCall(contract util.Uint160, method string, attrs []transaction.Attribute, params ...interface{}) (*transaction.Transaction, error)
+	SendCall(contract util.Uint160, method string, params ...interface{}) (util.Uint256, uint32, error)
+}
+
+// Hash stores the hash of the native RoleManagement contract.
+var Hash = state.CreateNativeContractHash(nativenames.Designation)
+
+const designateMethod = "designateAsRole"
+
+// ContractReader provides an interface to call read-only RoleManagement
+// contract's methods.
+type ContractReader struct {
+	invoker Invoker
+}
+
+// Contract represents a RoleManagement contract client that can be used to
+// invoke all of its methods.
+type Contract struct {
+	ContractReader
+
+	actor Actor
+}
+
+// DesignationEvent represents an event emitted by RoleManagement contract when
+// a new role designation is done.
+type DesignationEvent struct {
+	Role       noderoles.Role
+	BlockIndex uint32
+}
+
+// NewReader creates an instance of ContractReader that can be used to read
+// data from the contract.
+func NewReader(invoker Invoker) *ContractReader {
+	return &ContractReader{invoker}
+}
+
+// New creates an instance of Contract to perform actions using
+// using the given Actor. Notice that RoleManagement's state can be changed
+// only by the network's committee, so the Actor provided must be a committee
+// actor for designation methods to work properly.
+func New(actor Actor) *Contract {
+	return &Contract{*NewReader(actor), actor}
+}
+
+// GetDesignatedByRole returns the list of the keys designated to serve for the
+// given role at the given height. If no keys are configured for this role/height
+// combination nil is returned (with no error).
```

Comment:
Nitpicking, IIRC, an empty array is returned (not nil).

Reply:
See https://github.com/nspcc-dev/neo-go/blob/bc9b5d6976dbdb0aa5d95cb40b8a2ef8ccb4307a/pkg/core/native/designate.go#L292-L293 and https://github.com/nspcc-dev/neo-go/blob/bc9b5d6976dbdb0aa5d95cb40b8a2ef8ccb4307a/pkg/core/native/designate.go#L192-L196 and https://github.com/nspcc-dev/neo-go/blob/bc9b5d6976dbdb0aa5d95cb40b8a2ef8ccb4307a/pkg/core/native/native_neo.go#L1197-L1203

Reply:
> Nitpicking, IIRC, an empty array is returned (not nil).

Ah, yes, it was fixed in the code, but comment is outdated a bit now.

Reply:
Fixed.